modules/apps/dynamic-data-mapping/dynamic-data-mapping-test/src/testIntegration/java/com/liferay/dynamic/data/mapping/upgrade/v5_1_0/test/DDMFieldDDMFieldUpgradeProcessTest.java

```diff
@@ -0,0 +1,215 @@
+			serviceContext);
+
+		DDLRecordSet ddlRecordSet = _ddlRecordSetLocalService.addRecordSet(
+			TestPropsValues.getUserId(), _group.getGroupId(),
+			_ddmStructure.getStructureId(), null,
+			RandomTestUtil.randomLocaleStringMap(), null,
+			DDLRecordSetConstants.MIN_DISPLAY_ROWS_DEFAULT,
+			DDLRecordSetConstants.SCOPE_DYNAMIC_DATA_LISTS,
+			ServiceContextTestUtil.getServiceContext());
+
+		DDMFormValues ddmFormValues = _createDDMFormValues();
+
+		DDLRecord ddlRecord = _ddlRecordLocalService.addRecord(
+			TestPropsValues.getUserId(), _group.getGroupId(),
+			ddlRecordSet.getRecordSetId(),
+			DDLRecordConstants.DISPLAY_INDEX_DEFAULT, ddmFormValues,
+			serviceContext);
+
+		ddmStorageLink = ddmStorageLinkLocalService.getClassStorageLink(
+			ddlRecord.getDDMStorageId());
+
+		_priority = RandomTestUtil.randomInt();
+
+		_addDDMField("property", _priority);
+	}
+
+	@Override
+	protected CTModel<?> addCTModel() throws Exception {
+		return _addDDMField("content", _priority + 1);
+	}
+
+	@Override
+	protected CTService<?> getCTService() {
+		return _ddmFieldLocalService;
+	}
+
+	@Override
+	protected void runUpgrade() throws Exception {
+		UpgradeProcess upgradeProcess = UpgradeTestUtil.getUpgradeStep(
+			_upgradeStepRegistrator, _CLASS_NAME);
+
+		upgradeProcess.upgrade();
+	}
+
+	@Override
+	protected CTModel<?> updateCTModel(CTModel<?> ctModel) throws Exception {
+		DDMField ddmField = (DDMField)ctModel;
+
+		ddmField.setParentFieldId(RandomTestUtil.randomLong());
+
+		return _ddmFieldLocalService.updateDDMField(ddmField);
+	}
+
+	protected DDMStorageLink ddmStorageLink;
+
+	@Inject
+	protected DDMStorageLinkLocalService ddmStorageLinkLocalService;
+
+	private DDMField _addDDMField(String fieldName, int priority) {
+		DDMField ddmField = _ddmFieldLocalService.createDDMField(
+			_counterLocalService.increment());
+
+		ddmField.setParentFieldId(0);
+		ddmField.setStorageId(ddmStorageLink.getClassPK());
+		ddmField.setStructureVersionId(ddmStorageLink.getStructureVersionId());
+		ddmField.setFieldName(fieldName);
+		ddmField.setFieldType(DDMFormFieldTypeConstants.TEXT);
+		ddmField.setInstanceId(RandomTestUtil.randomString(8));
+		ddmField.setLocalizable(false);
+		ddmField.setPriority(priority);
+
+		return _ddmFieldLocalService.addDDMField(ddmField);
+	}
+
+	private DDMFormField _createDDMFormField(
+		String dataType, String name, String type) {
+
+		DDMFormField ddmFormField = new DDMFormField(name, type);
+
+		ddmFormField.setDataType(dataType);
+		ddmFormField.setLocalizable(true);
+
+		return ddmFormField;
+	}
+
+	private DDMFormValues _createDDMFormValues() {
+		DDMFormField ddmFormField = _createDDMFormField(
+			null, "parent", "fieldset");
+
+		ddmFormField.addNestedDDMFormField(
+			_createDDMFormField("string", "child", "text"));
+
+		DDMForm ddmForm = _ddmStructure.getDDMForm();
+
+		ddmForm.setDDMFormFields(ListUtil.fromArray(ddmFormField));
+
+		return new DDMFormValues(ddmForm);
```

Comment:
@p-albuquerque reorder:

```
DDMForm ddmForm = _ddmStructure.getDDMForm();

DDMFormField ddmFormField = DDMFormTestUtil.createDDMFormField(
	"parent", RandomTestUtil.randomString(),
	DDMFormFieldTypeConstants.FIELDSET, null, true, false, false);

ddmFormField.addNestedDDMFormField(
	DDMFormTestUtil.createDDMFormField(
		"child", RandomTestUtil.randomString(),
		DDMFormFieldTypeConstants.TEXT, "string", true, false, false));

ddmForm.setDDMFormFields(ListUtil.fromArray(ddmFormField));

return new DDMFormValues(ddmForm);
```

Reply:
@p-albuquerque please remember the "burrito" pattern (see example [here](https://github.com/liferay/liferay-portal/commit/f18bb2b3faa03754404047a5f5b0c277fa7e3d98))

In this case, `ddmFormField` was created to be used by `ddmForm`, so we first declare `ddmForm`, then create its dependencies, and then update `ddmForm`

```
parent

child

parent.set(child)

return parent
```